tests/s25Main/integration/testEconomyMode.cpp

```diff
@@ -0,0 +1,144 @@
+        inv.Add(goodsToCollect[0], amountsToAdd[playerIdx]);
+        world.GetSpecObj<nobBaseWarehouse>(hqPos[playerIdx])->AddGoods(inv, true);
+    }
+    world.GetEvMgr().ExecuteNextGF();
+    world.econHandler->UpdateAmounts();
+    for(unsigned playerIdx = 0; playerIdx < 3; ++playerIdx)
+    {
+        BOOST_REQUIRE_EQUAL(world.econHandler->GetAmount(0, playerIdx), initialAmount + amountsToAdd[playerIdx]);
+    }
+
+    // Teams behaviour
+    auto econTeams = world.econHandler->GetTeams();
+
+    // People get assigned to the correct teams?
+    BOOST_REQUIRE_EQUAL(econTeams.size(), (size_t)2);
+    unsigned smallTeam = econTeams[0].playersInTeam.count() - 1;
+    unsigned bigTeam = 2 - econTeams[0].playersInTeam.count();
+    BOOST_REQUIRE_EQUAL(econTeams[smallTeam].playersInTeam.count(), (size_t)1);
+    BOOST_REQUIRE_EQUAL(econTeams[bigTeam].playersInTeam.count(), (size_t)2);
+    BOOST_REQUIRE(econTeams[smallTeam].containsPlayer(1));
+    BOOST_REQUIRE(econTeams[bigTeam].containsPlayer(0));
+    BOOST_REQUIRE(econTeams[bigTeam].containsPlayer(2));
+
+    // Team amounts get calculated correctly?
+    BOOST_REQUIRE_EQUAL(econTeams[smallTeam].amountsTheTeamCollected[0], initialAmount + amountsToAdd[1]);
+    BOOST_REQUIRE_EQUAL(econTeams[bigTeam].amountsTheTeamCollected[0],
+                        2 * initialAmount + amountsToAdd[0] + amountsToAdd[2]);
+    BOOST_REQUIRE(econTeams[smallTeam].goodTypeWins < econTeams[bigTeam].goodTypeWins);
+    BOOST_REQUIRE_EQUAL(world.econHandler->GetMaxTeamAmount(0), econTeams[bigTeam].amountsTheTeamCollected[0]);
+
+    // Serialization/Deserialization
+    Savegame save;
+    for(unsigned i = 0; i < world.GetNumPlayers(); i++)
+        save.AddPlayer(world.GetPlayer(i));
+    save.ggs = ggs;
+    save.start_gf = em.GetCurrentGF();
+    save.sgd.MakeSnapshot(game);
+    TmpFile tmpFile;
+    BOOST_TEST_REQUIRE(save.Save(tmpFile.filePath, "MapTitle"));
+    Savegame loadSave;
+    BOOST_TEST_REQUIRE(loadSave.Load(tmpFile.filePath, SaveGameDataToLoad::All));
+    BOOST_TEST_REQUIRE(loadSave.GetNumPlayers() == (unsigned)3);
+    std::vector<PlayerInfo> players;
+    for(unsigned j = 0; j < 3; j++)
+        players.push_back(PlayerInfo(loadSave.GetPlayer(j)));
+    std::shared_ptr<Game> sharedGame(new Game(save.ggs, loadSave.start_gf, players));
+    GameWorld& newWorld = sharedGame->world_;
+    MockLocalGameState localGameState;
+    save.sgd.ReadSnapshot(sharedGame, localGameState);
+    BOOST_TEST_REQUIRE(newWorld.econHandler);
+    auto& goodsToCollectAfter = newWorld.econHandler->GetGoodTypesToCollect();
+    auto& econTeamsAfter = newWorld.econHandler->GetTeams();
+    BOOST_REQUIRE(goodsToCollect == goodsToCollectAfter);
+    BOOST_REQUIRE_EQUAL(econTeams.size(), econTeamsAfter.size());
+    newWorld.GetEvMgr().ExecuteNextGF();
+    newWorld.econHandler->UpdateAmounts();
+    for(auto& teamPair : boost::combine(econTeams, econTeamsAfter))
+    {
+        auto& before = teamPair.get_head();
```

Comment:
Using `boost::combine` is pretty neat! Maybe use `boost::tie(f, s) = var;` instead of the `get_head` stuff? Especially the `teamPair.get_tail().get_head();` disturbs me, I couldn't even tell if that is correct. Or maybe `get<0>(teamPair)` already works (`tie` won't work with references)?

Also please use `const auto&` (const wherever possible to avoid accidentally changing something) and be careful with `auto`: It makes code harder to read when the type is not clear, so I tend to use it only for hard-to-type types (e.g. iterators) or when the type appears on the RHS (`auto x = static_cast<Foo>(bar)`)